src/misc/rwqueue.cpp

```diff
@@ -74,6 +74,53 @@ void RWQueue<T>::Enqueue(T&& item)
 	has_items.notify_one();
 }
 
+template <typename T>
+void RWQueue<T>::BulkEnqueue(std::vector<T>& from_source, const size_t num_requested)
+{
+	// Check that the source can provide the quantity requested
+	assert(num_requested <= from_source.size());
+
+	// The source iterator will be incremented with each chunk enqueued
+	auto source_start = from_source.begin();
+
+	// The remaining count is decremented until no more items need enqueing
+	auto num_remaining = num_requested;
+
+	while (num_remaining > 0) {
+		// Begin exclusive access to the internal queue
+		std::unique_lock<std::mutex> lock(mutex);
+
+		// Limit the quantity to our capacity
+		const auto min_items = std::max(static_cast<size_t>(1),
+		                                static_cast<size_t>(capacity - queue.size()));
+		const auto num_items = std::min(num_remaining, min_items);
+
+		// Set the range of the source items we'll be working with
+		const auto stride     = static_cast<stride_t>(num_items);
+		const auto source_end = source_start + stride;
+
+		// wait until the queue has enough room for the items
+		has_room.wait(lock, [&] { return capacity - queue.size() >= num_items; });
+
+		// Enqueue the range of items from the source vector
+		queue.insert(queue.end(),
+		             std::move_iterator(source_start),
+		             std::move_iterator(source_end));
+
+		// notify the next waiting thread that we have an item
+		lock.unlock();
+		has_items.notify_one();
+
+		// Step the start forward to the prior end
+		source_start = source_end;
+
+		// Deduct the quantity done this pass
+		num_remaining -= num_items;
+	}
+	// clear the hollowed-out source vector so it can be re-used
+	from_source.clear();
```

Comment:
Can we not do this here and let the caller do it? I think it's better to avoid mutating passed-in data unless absolutely necessary (and side-effects in general).

Reply:
We're hollowing out the vector's contents (via std::move), so the mutation is happening (brundle :fly: !)

And the spec says that once a vector has elements moved out, its in an undefined state (similar to deleting content behind a pointer).

The `.clear()` is good house-keeping to prevent UBSAN, similar to assign `nullptr` to an invalidated pointer.

There's a good disussion here: https://stackoverflow.com/questions/52697122/is-it-necessary-to-reset-stdlist-after-been-moved

We could defer it, but it's just more defensive and cleaner to always clear() right after the point of std::move().

Reply:
I'll add more comments about this to the function.

Reply:
Added.

Reply:
Ah, it all makes sense then, I wasn't aware of all these little details :)